program/src/big_vec.rs

```diff
@@ -0,0 +1,490 @@
+
+        // if not last element replace it with last
+        if index != self.len() - 1 {
+            unsafe {
+                sol_memmove(
+                    self.data[start_index..end_index].as_mut_ptr(),
+                    self.data[data_end_index - T::LEN..data_end_index].as_mut_ptr(),
+                    T::LEN,
+                );
+            }
+        }
+
+        let new_len = self.len() - 1;
+        let mut vec_len_ref = &mut self.data[0..VEC_SIZE_BYTES];
+        new_len.serialize(&mut vec_len_ref)?;
+
+        Ok(value)
+    }
+
+    /// Extracts a slice of the data types
+    pub fn deserialize_mut_slice<T: Pack>(
+        &mut self,
+        skip: usize,
+        len: usize,
+    ) -> Result<Vec<&'data mut T>, ProgramError> {
+        let vec_len = self.len();
+        let last_item_index = skip
+            .checked_add(len)
+            .ok_or(ProgramError::AccountDataTooSmall)?;
+        if last_item_index > vec_len as usize {
+            return Err(ProgramError::AccountDataTooSmall);
+        }
+
+        let start_index = VEC_SIZE_BYTES.saturating_add(skip.saturating_mul(T::LEN));
+        let end_index = start_index.saturating_add(len.saturating_mul(T::LEN));
+        let mut deserialized = vec![];
+        for slice in self.data[start_index..end_index].chunks_exact_mut(T::LEN) {
+            deserialized.push(unsafe { &mut *(slice.as_ptr() as *mut T) });
+        }
+        Ok(deserialized)
+    }
+
+    /// Add new element to the end
+    pub fn push<T: Pack>(&mut self, element: T) -> Result<(), ProgramError> {
+        let mut vec_len_ref = &mut self.data[0..VEC_SIZE_BYTES];
+        let mut vec_len = u32::try_from_slice(vec_len_ref)?;
+
+        let start_index = VEC_SIZE_BYTES + vec_len as usize * T::LEN;
+        let end_index = start_index + T::LEN;
+
+        vec_len += 1;
+        vec_len.serialize(&mut vec_len_ref)?;
+
+        if self.data.len() < end_index {
+            return Err(ProgramError::AccountDataTooSmall);
+        }
+        let element_ref = &mut self.data[start_index..end_index];
+        element.pack_into_slice(element_ref);
+        Ok(())
+    }
+
+    /// Get an iterator for the type provided
+    pub fn iter<'vec, T: Pack>(&'vec self) -> Iter<'data, 'vec, T> {
+        Iter {
+            len: self.len() as usize,
+            current: 0,
+            current_index: VEC_SIZE_BYTES,
+            inner: self,
+            phantom: PhantomData,
+        }
+    }
+
+    /// Get a mutable iterator for the type provided
+    pub fn iter_mut<'vec, T: Pack>(&'vec mut self) -> IterMut<'data, 'vec, T> {
+        IterMut {
+            len: self.len() as usize,
+            current: 0,
+            current_index: VEC_SIZE_BYTES,
+            inner: self,
+            phantom: PhantomData,
+        }
+    }
+
+    /// Find matching data in the array
+    pub fn find<T: Pack>(&self, data: &[u8], predicate: fn(&[u8], &[u8]) -> bool) -> Option<&T> {
```

Comment:
Not quite, `Iterator::find` transmutes every value to `T` and compares to pubkey. Find with `predicate` will just memcmp first 32 bytes of a string with pubkey and transmute once on return. We eliminate almost all transmuting this way. Don't know if this gives a lot of performance boost, should benchmark it, but it will give some definitely.

UPDATE:
I benched it. I was able to handle only 4000 of validators vs 6700 previously. So we should leave it as is.